[Merged by Bors] - feat(topology/instances/rat): some facts about topology on `rat`

urkud opened the pull request:
* `ℚ` is a totally disconnected space;
* `cocompact  ℚ` is not a countably generated filter;
* hence, `alexandroff  ℚ` is not a first countable topological space.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

urkud committed:
Snapshot

urkud committed:
Drop duplicate

urkud committed:
Snapshot

urkud committed:
Update

vihdzp commented on src/data/fintype/basic.lean:
```diff
@@ -1775,37 +1775,63 @@ lemma of_injective [infinite β] (f : β → α) (hf : injective f) : infinite 
+instance [infinite α] : infinite (option α) :=
+infinite.of_injective some (option.some_injective α)
+
+instance sum.infinite_of_left [infinite α] : infinite (α ⊕ β) :=
```
Can't this and `sum.infinite_of_right` cause diamonds? They should probably be `def` instead.

urkud replied:
No, they're in `Prop`.

vihdzp commented on src/data/fintype/basic.lean:
```diff
@@ -1775,37 +1775,63 @@ lemma of_injective [infinite β] (f : β → α) (hf : injective f) : infinite 
 
-instance sum_of_right [infinite β] : infinite (α ⊕ β) :=
-of_injective sum.inr sum.inr_injective
+instance prod.infinite_of_right [nonempty α] [infinite β] : infinite (α × β) :=
```
Same here, this seems diamond-prone.

urkud committed:
Merge branch 'master' into YK-rat-cocompact

urkud committed:
Merge branch 'master' into YK-rat-cocompact

urkud committed:
refactor(topology/constructions): turn `cofinite_topology` into a type synonym

Instead of `cofinite_topology α : topological_space α`, define
`cofinite_topology α := α` with an instance
`topological_space (cofinite_topology α) := (old definition)`.

This way we can talk about cofinite topology without using `@` all
over the place.

Also move `homeo_of_equiv_compact_to_t2.t1_counterexample` to
`topology.alexandroff` and prove it for `alexandroff ℕ` and
`cofinite_topology (alexandroff ℕ)`.

urkud committed:
Apply suggestions from code review

urkud committed:
Merge branch 'YK-cofinite-topology' into YK-rat-cocompact

urkud committed:
Fix

urkud committed:
Merge branch 'YK-cofinite-topology' into YK-rat-cocompact

urkud committed:
Merge branch 'master' into YK-rat-cocompact

robertylewis commented on src/topology/instances/rat.lean:
```diff
@@ -0,0 +1,88 @@
+
+open set metric filter topological_space
+open_locale topological_space alexandroff
+local notation `ℚ∞` := alexandroff ℚ
```
Should this be added to the `alexandroff` locale, or do you think that's not helpful?

urkud replied:
I don't think that we'll use `alexandroff ℚ` for any purpose other than "build a strange counterexample".

robertylewis reviewed (approved):
bors d+
